File: engines/ios/Tests/TrajectoryRuntimeEngineTests/ConformanceRunner.swift
import XCTest
import Foundation
@testable import TrajectoryRuntimeEngine

final class ConformanceRunner: XCTestCase {

    private func findConformanceDir() -> URL {
        // Try relative to current working directory
        let cwd = FileManager.default.currentDirectoryPath
        let candidates = [
            URL(fileURLWithPath: cwd).appendingPathComponent("../../spec/conformance"),
            URL(fileURLWithPath: cwd).appendingPathComponent("../../../spec/conformance"),
            URL(fileURLWithPath: cwd).appendingPathComponent("spec/conformance"),
        ]

        for candidate in candidates {
            let standardized = candidate.standardized
            var isDir: ObjCBool = false
            if FileManager.default.fileExists(atPath: standardized.path, isDirectory: &isDir), isDir.boolValue {
                return standardized
            }
        }

        // Try Bundle resources (when tests copy conformance/)
        if let resourceURL = Bundle.module.resourceURL?.appendingPathComponent("conformance") {
            var isDir: ObjCBool = false
            if FileManager.default.fileExists(atPath: resourceURL.path, isDirectory: &isDir), isDir.boolValue {
                return resourceURL
            }
        }

        fatalError("Cannot find conformance directory")
    }

    private func discoverFixtures(baseDir: URL) -> [TestFixture] {
        var fixtures: [TestFixture] = []
        let subdirs = ["validation", "execution", "parameters"]
        let decoder = JSONDecoder()

        for subdir in subdirs {
            let dir = baseDir.appendingPathComponent(subdir)
            guard let files = try? FileManager.default.contentsOfDirectory(at: dir, includingPropertiesForKeys: nil) else {
                continue
            }
            let jsonFiles = files.filter { $0.pathExtension == "json" }.sorted { $0.lastPathComponent < $1.lastPathComponent }
            for file in jsonFiles {
                guard let data = try? Data(contentsOf: file),
                      let fixture = try? decoder.decode(TestFixture.self, from: data) else {
                    continue
                }
                fixtures.append(fixture)
            }
        }
        return fixtures
    }

    func testConformance() throws {
        let conformanceDir = findConformanceDir()
        let fixtures = discoverFixtures(baseDir: conformanceDir)

        XCTAssertGreaterThan(fixtures.count, 0, "No fixtures found in \(conformanceDir.path)")

        var passed = 0
        var failed = 0
        var failures: [String] = []

        for fixture in fixtures {
            do {
                try runFixture(fixture)
                passed += 1
            } catch {
                failed += 1
                failures.append("\(fixture.test_id): \(fixture.name) - \(error)")
            }
        }

        if !failures.isEmpty {
            let failureList = failures.joined(separator: "\n  ")
            XCTFail("\(failed) fixture(s) failed:\n  \(failureList)")
        }

        print("\(passed) passed, \(failed) failed, \(passed + failed) total")
    }

    private func runFixture(_ fixture: TestFixture) throws {
        if fixture.category == "validation" {
            try runValidationFixture(fixture)
        } else {
            try runExecutionFixture(fixture)
        }
    }

    private func runValidationFixture(_ fixture: TestFixture) throws {
        let workflowMap = fixture.workflow.toAny() as! [String: Any]
        let result = validate(workflow: workflowMap)

        if fixture.expected.valid != result.valid {
            let extra = result.error_code.map { " (error: \($0))" } ?? ""
            throw FixtureError.mismatch("Expected valid=\(fixture.expected.valid), got valid=\(result.valid)\(extra)")
        }

        if !fixture.expected.valid, let expectedCode = fixture.expected.error_code {
            if result.error_code != expectedCode {
                throw FixtureError.mismatch("Expected error_code=\"\(expectedCode)\", got \"\(result.error_code ?? "nil")\"")
            }
        }
    }

    private func runExecutionFixture(_ fixture: TestFixture) throws {
        // First validate
        let workflowMap = fixture.workflow.toAny() as! [String: Any]
        let valResult = validate(workflow: workflowMap)
        if !valResult.valid {
            throw FixtureError.mismatch("Validation failed: \(valResult.error_code ?? "unknown")")
        }

        // Deserialize workflow as typed object
        let decoder = JSONDecoder()
        let workflowJSON = try JSONEncoder().encode(fixture.workflow)
        let workflow = try decoder.decode(MasterWorkflowSpecification.self, from: workflowJSON)

        let engine = WorkflowEngine(workflow: workflow, setup: fixture.setup)
        engine.start()

        // Submit user actions
        if let actions = fixture.user_actions {
            for (i, action) in actions.enumerated() {
                engine.submitAction(action, actionIndex: i)
            }
        }

        // Compare execution trace
        if let expectedTrace = fixture.expected.execution_trace {
            let actualTrace = engine.getTrace()
            try compareTrace(expected: expectedTrace, actual: actualTrace)
        }

        // Compare workflow state
        if let expectedState = fixture.expected.workflow_state {
            let actualState = engine.getWorkflowState().rawValue
            if actualState != expectedState {
                throw FixtureError.mismatch("Workflow state: expected \"\(expectedState)\", got \"\(actualState)\"")
            }
        }

        // Compare final properties
        if let expectedProps = fixture.expected.final_properties {
            let actualProps = engine.getProperties()
            try compareProperties(expected: expectedProps, actual: actualProps)
        }
    }

    private func compareTrace(expected: [TraceEntry], actual: [TraceEntry]) throws {
        if expected.count != actual.count {
            let expectedList = expected.map { "\($0.step_oid):\($0.state)" }
            let actualList = actual.map { "\($0.step_oid):\($0.state)" }
            throw FixtureError.mismatch(
                "Trace length mismatch: expected \(expected.count), got \(actual.count)\n" +
                "  Expected: \(expectedList)\n" +
                "  Actual:   \(actualList)"
            )
        }

        for i in 0..<expected.count {
            let exp = expected[i]
            let act = actual[i]

            if exp.step_oid != act.step_oid {
                throw FixtureError.mismatch("Trace[\(i)] step_oid: expected \"\(exp.step_oid)\", got \"\(act.step_oid)\"")
            }
            if exp.state != act.state {
                throw FixtureError.mismatch("Trace[\(i)] state: expected \"\(exp.state)\", got \"\(act.state)\" (step \(exp.step_oid))")
            }
            if exp.order != act.order {
                throw FixtureError.mismatch("Trace[\(i)] order: expected \(exp.order), got \(act.order) (step \(exp.step_oid))")
            }
            if let expAction = exp.after_action, expAction != act.after_action {
                throw FixtureError.mismatch("Trace[\(i)] after_action: expected \(expAction), got \(act.after_action ?? -1) (step \(exp.step_oid))")
            }
        }
    }

    private func compareProperties(expected: [String: String], actual: [String: String]) throws {
        for (key, expectedValue) in expected {
            let actualValue = actual[key]
            if actualValue != expectedValue {
                throw FixtureError.mismatch("Property \"\(key)\": expected \"\(expectedValue)\", got \"\(actualValue ?? "(undefined)")\"")
            }
        }
    }

    enum FixtureError: Error, CustomStringConvertible {
        case mismatch(String)
        var description: String {
            switch self {
            case .mismatch(let msg): return msg
            }
        }
    }
}

File: engines/kmp-engine/src/jvmTest/kotlin/com/trajectoryruntime/engine/ConformanceRunner.kt
// Copyright (c) 2026 Saturnis.io. All rights reserved.
// Licensed under the GNU AGPL v3. See LICENSE.md for details.
package com.trajectoryruntime.engine

import com.fasterxml.jackson.databind.ObjectMapper
import kotlinx.serialization.json.Json
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory
import java.io.File

class ConformanceRunner {

    private val json = Json { ignoreUnknownKeys = true }
    private val mapper = ObjectMapper()

    private fun findConformanceDir(): File {
        val userDir = System.getProperty("user.dir") ?: "."
        val candidates = listOf(
            File(userDir, "../../spec/conformance"),
            File(userDir, "../../../spec/conformance"),
            File(userDir, "spec/conformance"),
        )
        for (dir in candidates) {
            if (dir.isDirectory) return dir.canonicalFile
        }
        throw IllegalStateException("Cannot find conformance directory, tried: ${candidates.map { it.absolutePath }}")
    }

    private fun discoverFixtures(baseDir: File): List<TestFixture> {
        val fixtures = mutableListOf<TestFixture>()
        val subdirs = listOf("validation", "execution", "parameters", "resources")

        for (subdir in subdirs) {
            val dir = File(baseDir, subdir)
            if (!dir.isDirectory) continue

            val files = dir.listFiles()
                ?.filter { it.extension == "json" }
                ?.sortedBy { it.name }
                ?: continue

            for (file in files) {
                val content = file.readText()
                val fixture = json.decodeFromString<TestFixture>(content)
                fixtures.add(fixture)
            }
        }

        return fixtures
    }

    @TestFactory
    fun conformanceTests(): List<DynamicTest> {
        val conformanceDir = findConformanceDir()
        val fixtures = discoverFixtures(conformanceDir)

        return fixtures.map { fixture ->
            DynamicTest.dynamicTest("${fixture.test_id}: ${fixture.name}") {
                runFixture(fixture)
            }
        }
    }

    private fun runFixture(fixture: TestFixture) {
        if (fixture.category == "validation") {
            runValidationFixture(fixture)
        } else {
            runExecutionFixture(fixture)
        }
    }

    private fun runValidationFixture(fixture: TestFixture) {
        // Convert JsonObject to Map<String, Any?> for the validator
        val workflowMap = jsonObjectToMap(fixture.workflow)
        val result = validate(workflowMap)

        if (fixture.expected.valid != result.valid) {
            val extra = if (result.error_code != null) " (error: ${result.error_code})" else ""
            throw AssertionError(
                "Expected valid=${fixture.expected.valid}, got valid=${result.valid}$extra"
            )
        }

        if (!fixture.expected.valid && fixture.expected.error_code != null) {
            if (result.error_code != fixture.expected.error_code) {
                throw AssertionError(
                    "Expected error_code=\"${fixture.expected.error_code}\", got \"${result.error_code}\""
                )
            }
        }
    }

    private fun runExecutionFixture(fixture: TestFixture) {
        // First validate
        val workflowMap = jsonObjectToMap(fixture.workflow)
        val valResult = validate(workflowMap)
        if (!valResult.valid) {
            throw AssertionError("Validation failed: ${valResult.error_code}")
        }

        // Deserialize workflow as typed object
        val workflowJson = fixture.workflow.toString()
        val workflow = json.decodeFromString<MasterWorkflowSpecification>(workflowJson)
        val engine = WorkflowEngine(workflow, fixture.setup)
        engine.start()

        // Submit user actions
        fixture.user_actions?.forEachIndexed { i, action ->
            engine.submitAction(action, i)
        }

        // Compare execution trace
        fixture.expected.execution_trace?.let { expectedTrace ->
            val actualTrace = engine.getTrace()
            compareTrace(expectedTrace, actualTrace)
        }

        // Compare workflow state
        fixture.expected.workflow_state?.let { expectedState ->
            val actualState = engine.getWorkflowState().name
            if (actualState != expectedState) {
                throw AssertionError(
                    "Workflow state: expected \"$expectedState\", got \"$actualState\""
                )
            }
        }

        // Compare final properties
        fixture.expected.final_properties?.let { expectedProps ->
            val actualProps = engine.getProperties()
            compareProperties(expectedProps, actualProps)
        }
    }

    private fun compareTrace(expected: List<TraceEntry>, actual: List<TraceEntry>) {
        if (expected.size != actual.size) {
            throw AssertionError(
                "Trace length mismatch: expected ${expected.size}, got ${actual.size}\n" +
                    "  Expected: ${expected.map { "${it.step_oid}:${it.state}" }}\n" +
                    "  Actual:   ${actual.map { "${it.step_oid}:${it.state}" }}"
            )
        }

        for (i in expected.indices) {
            val exp = expected[i]
            val act = actual[i]

            if (exp.step_oid != act.step_oid) {
                throw AssertionError("Trace[$i] step_oid: expected \"${exp.step_oid}\", got \"${act.step_oid}\"")
            }
            if (exp.state != act.state) {
                throw AssertionError("Trace[$i] state: expected \"${exp.state}\", got \"${act.state}\" (step ${exp.step_oid})")
            }
            if (exp.order != act.order) {
                throw AssertionError("Trace[$i] order: expected ${exp.order}, got ${act.order} (step ${exp.step_oid})")
            }
            if (exp.after_action != null && exp.after_action != act.after_action) {
                throw AssertionError("Trace[$i] after_action: expected ${exp.after_action}, got ${act.after_action} (step ${exp.step_oid})")
            }
        }
    }

    private fun compareProperties(expected: Map<String, String>, actual: Map<String, String>) {
        for ((key, expectedValue) in expected) {
            val actualValue = actual[key]
            if (actualValue != expectedValue) {
                throw AssertionError(
                    "Property \"$key\": expected \"$expectedValue\", got \"${actualValue ?: "(undefined)"}\""
                )
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun jsonObjectToMap(jsonObject: kotlinx.serialization.json.JsonObject): Map<String, Any?> {
        // Use Jackson to convert the JSON string to a Map
        val jsonString = jsonObject.toString()
        return mapper.readValue(jsonString, Map::class.java) as Map<String, Any?>
    }
}
